versioned/storage/common/src/main/java/org/projectnessie/versioned/storage/common/logic/ReferenceLogicImpl.java

```diff
@@ -0,0 +1,636 @@
+    }
+
+    enum Kind {
+      ADDED_TO_INDEX,
+      REF_ROW_EXISTS,
+      REF_ROW_MISSING
+    }
+  }
+
+  @VisibleForTesting // needed to simulate recovery scenarios
+  CommitReferenceResult commitCreateReference(String name, ObjId pointer)
+      throws RetryTimeoutException {
+    try {
+      return commitRetry(
+          persist,
+          (p, retryState) -> {
+            Reference refRefs = requireNonNull(p.fetchReference(REF_REFS.name()));
+            long created = p.config().currentTimeMicros();
+            RefObj ref = ref(name, pointer, created);
+            try {
+              p.storeObj(ref);
+            } catch (ObjTooLargeException e) {
+              throw new RuntimeException(e);
+            }
+
+            StoreKey k = key(name);
+
+            Instant now = persist.config().clock().instant();
+            CreateCommit c =
+                newCommitBuilder()
+                    .message("Create reference " + name + " pointing to " + pointer)
+                    .parentCommitId(refRefs.pointer())
+                    .addAdds(commitAdd(k, 0, requireNonNull(ref.id()), null, null))
+                    .headers(
+                        newCommitHeaders()
+                            .add("operation", "create")
+                            .add("name", name)
+                            .add("head", pointer.toString())
+                            .add("timestamp", now.toString())
+                            .add("timestamp.millis", Long.toString(now.toEpochMilli()))
+                            .build())
+                    .commitType(CommitType.INTERNAL)
+                    .build();
+
+            commitReferenceChange(p, refRefs, c);
+
+            return new CommitReferenceResult(reference(name, pointer, false), ADDED_TO_INDEX);
+          });
+    } catch (CommitConflictException e) {
+      checkState(e.conflicts().size() == 1, "Unexpected amount of conflicts %s", e.conflicts());
+
+      CommitConflict conflict = e.conflicts().get(0);
+      checkState(conflict.conflictType() == KEY_EXISTS, "Unexpected conflict type %s", conflict);
+
+      Supplier<StoreIndex<CommitOp>> indexSupplier = createRefsIndexSupplier();
+      StoreIndexElement<CommitOp> el = indexSupplier.get().get(key(name));
+      checkNotNull(el, "Key %s missing in index", name);
+
+      Reference existing = persist.fetchReference(name);
+
+      if (existing != null) {
+        return new CommitReferenceResult(existing, REF_ROW_EXISTS);
+      }
+
+      RefObj ref;
+      try {
+        ref =
+            persist.fetchTypedObj(
+                requireNonNull(el.content().value(), "Reference commit operation has no value"),
+                REF,
+                RefObj.class);
+      } catch (ObjNotFoundException ex) {
+        throw new RuntimeException("Internal error getting reference creation object", e);
+      }
+      return new CommitReferenceResult(
+          reference(name, ref.initialPointer(), false), REF_ROW_MISSING);
```

Comment:
It seems to me that moving the two "ref row *" sub-cases to `createReference` would make the overall logic easier to follow. This method would also contain the commit login, `createReference` would deal with finding and inserting other rows.  WDYT?

Reply:
Sounds rather cosmetic to me.